problem-solver/cxx/exampleModule/searcher/PathSearcher.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * This source file is part of an OSTIS project. For the latest info, see
+ * http://ostis.net Distributed under the MIT License (See accompanying file
+ * COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+ */
+
+#include "PathSearcher.hpp"
+
+#include "keynodes/Keynodes.hpp"
+#include "utils/NumberUtils.hpp"
+
+PathSearcher::PathSearcher(ScMemoryContext * context)
+  : context(context)
+{
+}
+
+void PathSearcher::findPath(
+    ScAddr const & graph,
+    ScAddr const & startNode,
+    ScAddr const & endNode,
+    ConnectorTemplateInfo const & connectorTemplateInfo,
+    WeightTemplateInfo const & weightTemplateInfo,
+    unsigned & pathLength,
+    ScAddrVector & path) const
+{
+  ScAddrQueue vertexesToCheck;
+  ScAddrToValueUnorderedMap<unsigned> pathLengthToVertexes;
+  pathLengthToVertexes[startNode] = 0;
+  ScAddrToValueUnorderedMap<ScAddrVector> pathsToVertexes;
+
+  vertexesToCheck.emplace(startNode);
+
+  while (!vertexesToCheck.empty())
+  {
+    ScAddr const & currentVertex = vertexesToCheck.front();
+    vertexesToCheck.pop();
+
+    ScAddrToValueUnorderedMap<unsigned> neighborsWithPathLength;
+    getNeighborsWithConnectorsLength(
+        graph, currentVertex, connectorTemplateInfo, weightTemplateInfo, neighborsWithPathLength);
+
+    for (auto const & neighborWithPathLength : neighborsWithPathLength)
+    {
+      ScAddr const & neighbor = neighborWithPathLength.first;
+      unsigned const connectorWeight = neighborWithPathLength.second;
+
+      if (pathLengthToVertexes.find(neighbor) == pathLengthToVertexes.cend())
+      {
+        pathLengthToVertexes[neighbor] = pathLengthToVertexes[currentVertex] + connectorWeight;
+
+        pathsToVertexes[neighbor] = pathsToVertexes[currentVertex];
+        pathsToVertexes[neighbor].emplace_back(neighbor);
+
+        vertexesToCheck.emplace(neighbor);
+      }
+      else
+      {
+        unsigned newPathLength = pathLengthToVertexes[currentVertex] + connectorWeight;
+        if (newPathLength < pathLengthToVertexes[neighbor])
+        {
+          pathLengthToVertexes[neighbor] = newPathLength;
+
+          pathsToVertexes[neighbor] = pathsToVertexes[currentVertex];
+          pathsToVertexes[neighbor].emplace_back(neighbor);
+
+          vertexesToCheck.emplace(neighbor);
+        }
```

Comment:
I removed comment about duplicated code here previous time because I thought that its logic will be completely changed, but these branches are still very much identical
something like callback inside getNeighborsWithConnectorsLength can be implemented here